metadataserver/authentication.go

```diff
@@ -0,0 +1,78 @@
+package metadataserver
+
+import (
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/Netflix/titus-executor/metadataserver/auth"
+	"golang.org/x/net/ipv4"
+)
+
+func (ms *MetadataServer) authenticate(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		token := r.Header.Get("X-aws-ec2-metadata-token")
+		if !ms.tokenRequired && len(token) == 0 {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		auth := auth.HMACAuthenticator{Key: ms.tokenKey}
+		if !auth.VerifyToken(token) {
+			http.Error(w, "", http.StatusUnauthorized)
```

Comment:
It might be worth indicating if the token is totally invalid vs. not present vs. expired

Reply:
This is the way IMDS v2 actually behaves. I’m concern we’d break clients if we deviate.

Reply:
Can we log the reason at least?

Reply:
All the logs end up in journald, and we can setup alerting based on it.

Reply:
Yep Yep, good cal. I'll do a logging pass through all of the new code.

Reply:
Yep yep, good call. I'll do a logging pass on all of my new code. I'll take a look at adding handy metrics as well.